osf/migrations/0171_populate_file_versions_through.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.15 on 2019-03-03 17:52
+from __future__ import unicode_literals
+
+import logging
+
+from django.db import migrations, connection
+
+logger = logging.getLogger(__file__)
+
+
+def restore_default_through_table(state, schema):
+    sql = """
+        DROP TABLE osf_basefilenode_versions;
+        CREATE TABLE osf_basefilenode_versions AS
+        SELECT
+            new_thru.basefilenode_id,
+            new_thru.fileversion_id
+        FROM
+            osf_basefileversionsthrough AS new_thru;
+
+        ALTER TABLE osf_basefilenode_versions ADD COLUMN id SERIAL PRIMARY KEY;
+        ALTER TABLE osf_basefilenode_versions ADD CONSTRAINT osf_basefilenod_basefilenode_id_b0knah27_fk_osf_basefilenode_id FOREIGN KEY (basefilenode_id) REFERENCES osf_basefilenode DEFERRABLE INITIALLY DEFERRED;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN basefilenode_id
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN fileversion_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN fileversion_id
+        SET
+            DATA TYPE INTEGER;
+        ALTER TABLE osf_basefilenode_versions ALTER COLUMN basefilenode_id
+        SET
+            NOT NULL;
+        ALTER TABLE osf_basefilenode_versions ADD CONSTRAINT osf_basefilenode__fileversion_id_93etanfc_fk_osf_fileversion_id FOREIGN KEY (fileversion_id) REFERENCES osf_fileversion DEFERRABLE INITIALLY DEFERRED;
+        ALTER TABLE osf_basefilenode_versions ADD CONSTRAINT osf_basefilenode__fileversion_uniq564 UNIQUE (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (basefilenode_id, fileversion_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (basefilenode_id);
+        CREATE INDEX
+        ON osf_basefilenode_versions (fileversion_id);
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(sql)
+
+
+def populate_fileversion_name(state, schema):
+
+    sql = """
+        DROP TABLE osf_basefileversionsthrough;
```

Comment:
Any benchmarking done regarding how long this will take to migrate ~7.5M rows?

If it's too long, an incremental approach might be better, where we migrate small amounts at a time via cron, while maintaining reads from the old table and writes to both, until we can fail over to using exclusively the new.

Alternatively, could we just rename the old table, have `version_name` nullable, and incrementally populate it?